Architecture diagrams (#119)
canonical/dns-operators: docs/explanation/charm-architecture.md

> # Charm architecture

Reviewer: ```suggestion
# Project architecture
```
These diagrams are great! They seem to be more about a high-level overview of the project though

  
  Component(tls-policy, "TLS policy")
  Component(dns-policy, "DNS policy")
  
  Rel(dns-policy, bind, "dns_record")
  UpdateRelStyle(dns-policy, bind, $textColor="green", $lineColor="green", $offsetX="-20", $offsetY="10")
  
  Component(lego, "Lego")
  Component(bind, "Bind charm")
  Component(httpreq, "HTTP Request Lego Provider")
  
  Rel(httpreq, bind, "dns_record")
  Rel(lego, httpreq, "API")
  Rel(tls-policy, lego, "tls-certificates")
  UpdateRelStyle(httpreq, bind, $textColor="green", $lineColor="green", $offsetX="10")
  UpdateRelStyle(lego, httpreq, $textColor="purple", $lineColor="purple", $offsetY="10")
  UpdateRelStyle(tls-policy, lego, $textColor="red", $lineColor="red")
}

Container_Boundary(app-model, "App model") {
    Component(app-1, "Application 1")
    Component(ingress, "Ingress")
    Component(app-2, "Application 2")

    Rel(app-1, ingress, "ingress")
    Rel(app-2, ingress, "ingress")
    UpdateRelStyle(app-1, ingress, $textColor="blue", $lineColor="blue", $offsetX="10")
    UpdateRelStyle(app-2, ingress, $textColor="blue", $lineColor="blue", $offsetX="10")
}

Rel(ingress, dns-policy, "dns_record")
Rel(ingress, tls-policy, "tls-certificates")

> UpdateRelStyle(ingress, dns-policy, $textColor="green", $lineColor="green")
> UpdateRelStyle(ingress, tls-policy, $textColor="red", $lineColor="red")
> 
> Person(operator, "Operator")
> Rel(operator, dns-policy, "approve/denies requests")
> UpdateRelStyle(ingress, dns-policy, $textColor="grey", $lineColor="grey", $offsetX="50")
> 
> Rel(operator, tls-policy, "approve/denies requests")
> UpdateRelStyle(ingress, tls-policy, $textColor="grey", $lineColor="grey", $offsetY="20")
> 
> UpdateRelStyle(operator, dns-policy, $offsetY="-40", $offsetX="-60")
> UpdateRelStyle(operator, tls-policy, $offsetY="-200", $offsetX="-155")

Reviewer: ```suggestion
UpdateRelStyle(ingress, dns-policy, $textColor="green", $lineColor="green", $offsetX="30", $offsetY="-10")
UpdateRelStyle(ingress, tls-policy, $textColor="red", $lineColor="red", $offsetX="-120", $offsetY="-10")

Person(operator, "Operator")
Rel(operator, dns-policy, "approve/denies requests")
Rel(operator, tls-policy, "approve/denies requests")
UpdateRelStyle(operator, dns-policy, $offsetY="-40", $offsetX="-60")
UpdateRelStyle(operator, tls-policy, $offsetY="-200", $offsetX="-155")
```

I did a bit of reorganizing because I saw that you had duplicates of the `UpdateRelStyle(ingress, dns-policy, ...)` and `UpdateRelStyle(ingress, tls-policy, ...)` that had conflicting style choices (green and red versus grey lines and text).

```

The bind-operator is usually deployed with dns-policy to enable human and/or automated approval of incoming DNS record requests.

> The workload of dns-policy is a Django application packaged as a snap with additional tooling. Since both the workloads of bind-operator and dns-policy

Reviewer: ```suggestion
The workload of dns-policy is a Django application packaged as a snap with additional tooling. Since the workloads of both bind-operator and dns-policy
```
Nitpick; this is a bit more understandable to me, but not a strong opinion

are snaps, they can work on the same machine. It was therefore decided to make dns-policy a subordinate charm.

The following diagram shows the interactions between bind-operator, dns-policy, and external components of a typical deployment of the DNS charms:

```mermaid
C4Container
title DNS charms components

UpdateLayoutConfig($c4ShapeInRow="3", $c4BoundaryInRow="2")

Rel(operator, charmed-dns-policy, "API", "approve/denies requests")

Component(postgresql, "Postgresql")
Person(operator, "Operator")

Component(empty1, "")
Component(empty2, "")
Component(empty3, "")
UpdateElementStyle(empty1,  $bgColor="#0000", $borderColor="#0000")
UpdateElementStyle(empty2,  $bgColor="#0000", $borderColor="#0000")
UpdateElementStyle(empty3,  $bgColor="#0000", $borderColor="#0000")

Component(dns-integrator, "DNS integrator")
Rel(dns-integrator, dns-policy, "dns_record")
UpdateRelStyle(dns-integrator, dns-policy, $textColor="green", $lineColor="green")

Rel(charmed-dns-policy, postgresql, "database")

> UpdateRelStyle(charmed-dns-policy, postgresql, $textColor="blue", $lineColor="blue")

Reviewer: ```suggestion
UpdateRelStyle(charmed-dns-policy, postgresql, $textColor="blue", $lineColor="blue", $offsetX="10")
```
Nitpick

Container_Boundary(machine-1, "Machine 1") { 

  Component(charmed-dns-policy, "DNS policy snap")

  Component(empty6, "")
  UpdateElementStyle(empty6,  $bgColor="#0000", $borderColor="#0000")

  Component(charmed-bind, "Bind snap")

  Component(empty0, "")
  UpdateElementStyle(empty0,  $bgColor="#0000", $borderColor="#0000")

  Component(empty5, "")
  UpdateElementStyle(empty5,  $bgColor="#0000", $borderColor="#0000")

  Component(empty7, "")
  UpdateElementStyle(empty7,  $bgColor="#0000", $borderColor="#0000")

  Component(dns-policy, "DNS policy charm")

  Component(empty8, "")
  UpdateElementStyle(empty8,  $bgColor="#0000", $borderColor="#0000")

  Component(bind, "Bind charm")

  Rel(dns-policy, bind, "dns_record")

>   UpdateRelStyle(dns-policy, bind, $textColor="green", $lineColor="green")

Reviewer: ```suggestion
  UpdateRelStyle(dns-policy, bind, $textColor="green", $lineColor="green", $offsetY="10")
```

  Rel(bind, charmed-bind, "API", "Update DNS records")

>   Rel(dns-policy, charmed-dns-policy, "API", "Update DNS requests")

Reviewer: ```suggestion
  Rel(dns-policy, charmed-dns-policy, "API", "Update DNS requests")
  UpdateRelStyle(bind, charmed-bind, $offsetX="10")
  UpdateRelStyle(dns-policy, charmed-dns-policy, $offsetX="10")
```

}
```

> After having deployed and related all the DNS charms in one model, you should see the following with `juju status --relations`:

Reviewer: ```suggestion
After having deployed and integrated all the DNS charms in one model, you should see the following output with `juju status --relations`:
```

0        started  10.227.61.237  juju-cd7a0a-0  ubuntu@22.04      Running
1        started  10.227.61.161  juju-cd7a0a-1  ubuntu@22.04      Running
2        started  10.227.61.146  juju-cd7a0a-2  ubuntu@22.04      Running
3        started  10.227.61.162  juju-cd7a0a-3  ubuntu@22.04      Running
4        started  10.227.61.3    juju-cd7a0a-4  ubuntu@22.04      Running

Integration provider            Requirer                            Interface          Type         Message
bind:bind-peers                 bind:bind-peers                     bind-instance      peer
bind:dns-record                 dns-policy:dns-record-requirer      dns_record         subordinate
dns-policy:dns-record-provider  dns-integrator-operator:dns-record  dns_record         regular
postgresql:database             dns-policy:database                 postgresql_client  regular
postgresql:database-peers       postgresql:database-peers           postgresql_peers   peer
postgresql:restart              postgresql:restart                  rolling_op         peer
postgresql:upgrade              postgresql:upgrade                  upgrade            peer
```

> You can here observe that the DNS policy charm is a subordinate one: it will be deployed on the same machine as the Bind charm as soon as it is related to it.

Reviewer: ```suggestion
You can here observe that the DNS policy charm is a subordinate one: It will be deployed on the same machine as the Bind charm as soon as it is integrated to it.
```

Author: Should "It" really be uppercase here ?

Author: TIL
